cdc/scheduler/internal/tp/table.go

```diff
@@ -0,0 +1,355 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tp
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/processor/pipeline"
+	"github.com/pingcap/tiflow/cdc/scheduler/internal"
+	"github.com/pingcap/tiflow/cdc/scheduler/internal/tp/schedulepb"
+	"go.uber.org/zap"
+)
+
+// table is a state machine that manage the table's state,
+// also tracking its progress by utilize the `TableExecutor`
+type table struct {
+	id model.TableID
+
+	status   schedulepb.TableStatus
+	executor internal.TableExecutor
+
+	task *dispatchTableTask
+}
+
+func newTable(tableID model.TableID, executor internal.TableExecutor) *table {
+	return &table{
+		id:       tableID,
+		executor: executor,
+		status:   schedulepb.TableStatus{TableID: tableID},
+		task:     nil,
+	}
+}
+
+// refresh the table' status, return true if the table state changed,
+// should be called before any table operations.
+func (t *table) refresh() bool {
+	oldState := t.status.State
+
+	meta := t.executor.GetTableMeta(t.id)
+	state := tableStatus2PB(meta.State)
+
+	t.status.State = state
+	t.status.Checkpoint.CheckpointTs = meta.CheckpointTs
+	t.status.Checkpoint.ResolvedTs = meta.ResolvedTs
+
+	if oldState != state {
+		log.Info("tpscheduler: table state changed",
+			zap.Any("oldState", oldState), zap.Any("state", state))
+		return true
+	}
+
+	return false
+}
+
+func newAddTableResponseMessage(status schedulepb.TableStatus) *schedulepb.Message {
+	return &schedulepb.Message{
+		MsgType: schedulepb.MsgDispatchTableResponse,
+		DispatchTableResponse: &schedulepb.DispatchTableResponse{
+			Response: &schedulepb.DispatchTableResponse_AddTable{
+				AddTable: &schedulepb.AddTableResponse{
+					Status:     &status,
+					Checkpoint: status.Checkpoint,
+				},
+			},
+		},
+	}
+}
+
+func newRemoveTableResponseMessage(status schedulepb.TableStatus) *schedulepb.Message {
+	message := &schedulepb.Message{
+		MsgType: schedulepb.MsgDispatchTableResponse,
+		DispatchTableResponse: &schedulepb.DispatchTableResponse{
+			Response: &schedulepb.DispatchTableResponse_RemoveTable{
+				RemoveTable: &schedulepb.RemoveTableResponse{
+					Status:     &status,
```

Comment:
Will the `Status` be a pointer to the `table.Status`? Will there be a data race?

Reply:
no, this is no data race.